src/lib/concurrency/logging/binary_recovery.cpp

```diff
@@ -0,0 +1,156 @@
+#include "binary_recovery.hpp"
+
+#include <fstream>
+
+#include "../../operators/insert.hpp"
+#include "../../storage/storage_manager.hpp"
+#include "../../storage/table.hpp"
+#include "../transaction_manager.hpp"
+#include "logger.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+BinaryRecovery& BinaryRecovery::getInstance() {
+  static BinaryRecovery instance;
+  return instance;
+}
+
+template <>
+std::string BinaryRecovery::_read(std::ifstream& file) {
+  std::string result;
+  std::getline(file, result, '\0');
+  return result;
+}
+
+AllTypeVariant BinaryRecovery::_read_AllTypeVariant(std::ifstream& file, DataType data_type) {
+  AllTypeVariant value;
+  switch (data_type) {
+    case DataType::Int: {
+      return AllTypeVariant{_read<int32_t>(file)};
+      break;
+    }
+    case DataType::Long: {
+      return AllTypeVariant{_read<int64_t>(file)};
+      break;
+    }
+    case DataType::Float: {
+      return AllTypeVariant{_read<float>(file)};
+      break;
+    }
+    case DataType::Double: {
+      return AllTypeVariant{_read<double>(file)};
+      break;
+    }
+    case DataType::String: {
+      return AllTypeVariant{_read<std::string>(file)};
+      break;
+    }
+    default:
+      DebugAssert(false, "recovery: read unknown type");
+  }
+
+  return value;
+}
+
+void BinaryRecovery::recover() {
+  TransactionID last_transaction_id{0};
+  for (auto& log_path : Logger::get_all_log_file_paths()) {
+    std::ifstream log_file(log_path);
+    DebugAssert(log_file.is_open(), "Recovery: could not open logfile " + log_path);
+
+    std::vector<LoggedItem> transactions;
+
+    while (true) {
+      // Begin of all Entries:
+      //   - log entry type       : sizeof(char)
+      //   - transaction_id       : sizeof(TransactionID)
+
+      char log_type;
+      log_file.read(&log_type, sizeof(char));
+
+      if (log_file.eof()) {
+        break;
+      }
+
+      DebugAssert(log_type == 't' || log_type == 'i' || log_type == 'v' || log_type == 'l',
+                  "Recovery: invalid log type token");
+
+      // if load entry
+      if (log_type == 'l') {
```

Comment:
switch? That would also take care unhandled values (via default) and fallthroughs.